src/Symfony/Component/Console/Command/CompleteCommand.php

```diff
@@ -0,0 +1,136 @@
+    ];
+    private $logger;
+
+    public function __construct(LoggerInterface $logger = null)
+    {
+        $this->logger = $logger;
+
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setName(ltrim(self::$defaultName, '|'))
+            ->setDescription(self::$defaultDescription)
+            ->setHidden(true)
+            ->addOption('shell', 's', InputOption::VALUE_REQUIRED)
+            ->addOption('current', 'c', InputOption::VALUE_REQUIRED)
+            ->addOption('input', 'i', InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY)
+        ;
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        try {
+            $shell = $input->getOption('shell');
+            if (!$shell) {
+                throw new \RuntimeException('The "--shell" option must be set.');
+            }
+
+            if (!$completionOutput = self::$completionOutputs[$shell] ?? false) {
+                throw new \RuntimeException(sprintf('Shell completion is not supported for your shell: "%s" (supported: "%s").', $shell, implode('", "', array_keys(self::$completionOutputs))));
+            }
+
+            $completionInput = $this->createCompletionInput($input);
+            $completion = new Completion();
+
+            $command = $this->findCommand($completionInput, $output);
+            if (null === $command) {
+                $this->getApplication()->complete($completionInput, $completion);
+            } elseif (
+                CompletionInput::TYPE_ARGUMENT_VALUE === $completionInput->getCompletionType()
+                && 'command' === $completionInput->getCompletionName()
+                && $command->getName() !== $completionInput->getCompletionValue()
+            ) {
+                $completion->suggestValue($command->getName());
+            } else {
+                $command->mergeApplicationDefinition();
+                $completionInput->bind($command->getDefinition());
+
+                $command->complete($completionInput, $completion);
+            }
+
+            $completionOutput = new $completionOutput();
+            $completionOutput->write($completion, $output);
+        } catch (\Throwable $e) {
+            if ($this->logger) {
+                $this->logger->warning('Error occurred while running shell completion for "{input}".', ['input' => (string) ($completionInput ?? $input), 'exception' => $e]);
+            }
+
+            if ($output->isVerbose()) {
+                throw $e;
+            }
+
+            return self::FAILURE;
+        }
+
+        return self::SUCCESS;
+    }
+
+    private function createCompletionInput(InputInterface $input): CompletionInput
+    {
+        $currentIndex = $input->getOption('current');
+        $completionInput = CompletionInput::fromTokens(array_map(
+            function (string $i): string { return trim($i, "'"); },
+            $input->getOption('input')
```

Comment:
I've noted these need to be validated better in the comments.

Reply:
Are gout maybe able to show me how to produce the errors you've shown? I haven't received those errors while writing/testing this

Reply:
Sure, try

```
bin/console _complete -v -s bash

bin/console _complete -v -s bash -c 1

in/console _complete -v -s bash -c 2 -i a

bin/console _complete -v -s bash -c a -i 2
```